frontend/controller/actions/contract.js

```diff
@@ -0,0 +1,28 @@
+import sbp from '~/shared/sbp.js'
+import { CONTRACT_IS_SYNCING } from '@utils/events.js'
+
+export default sbp('sbp/selectors/register', {
+  'gi.actions/contract/sync': async function (contractIDs: string | array) {
+    const listOfIds = typeof contractIDs === 'string' ? [contractIDs] : contractIDs
+
+    for (const id of listOfIds) {
+      await sbp('state/enqueueContractSync', id)
+    }
+  },
+  'gi.actions/contract/syncAndWait': async function (contractIDs: string | array) {
+    const listOfIds = typeof contractIDs === 'string' ? [contractIDs] : contractIDs
+
+    const promises = listOfIds.map(id => new Promise((resolve, reject) => {
+      const isDoneSyncing = (contractID, isSyncing) => {
+        if (isSyncing === false && contractID === id) {
+          sbp('okTurtles.events/off', CONTRACT_IS_SYNCING, isDoneSyncing)
+          resolve(id)
+        }
+      }
+      sbp('okTurtles.events/on', CONTRACT_IS_SYNCING, isDoneSyncing)
+      sbp('state/enqueueContractSync', id)
+    }))
```

Comment:
@taoeffect I think there was a bug on your Promise solution and I fixed it. The bug fix is we should enqueueContractSync inside the promise, not before. Otherwise, the contract could finish syncing even before creating the promise, and then the promise would never resolve. That bug happened on signup, but on group creation it didn't happen. Why?


Still... this solution didn't fix our core problem here: BypassUI (and any other action) is still returning (resolving) too soon. We must wait for the server to sync and that's not happening dunno why :/

Reply:
@sandrina-p there are two bugs in this code that work together to break things:

1. This function is labeled `async` even though it shouldn't be. Remove the `async` qualifier since that should only ever be used when there is a corresponding `await` (frankly I don't understand why our linter didn't complain)
2. Line 26 below needs to read `return Promise.all(promises)`. Right now a promise is being returned (because of the `async` qualifier), but not the one we want

Reply:
hum... but it is an asynchronous function because of Promise. I [followed this tutorial](https://www.taniarascia.com/promise-all-with-async-await/). And if I remove it, then the signup stops working. Maybe it would be better you could checkout this branch and see it locally

Reply:
> And if I remove it, then the signup stops working. 

You must do both things, remove it and fix line 26, otherwise it won't work. The rule is that `async` functions must have `await`s in them. If they do not, then the function must not have an `async` on it. Return a promise and you can still call await on it. You'll notice everything I'm saying here is validated by that tutorial.

Reply:
Writing `async` is equivalent to wrapping everything in the function in a promise and returning that promise. So by having `async` here the wrong promise is being returned. It needs to be the one on line 26

Reply:
Hum... still, this isn't working and I'm struggling to understand why :/
When we call `/signupAndLogin` on signup modal, the 2nd call `/login` inside it isn't resolved, preventing the form to finalize the submit (in this case, to close the modal).

Based on the logs, it seems we listen for `CONTRACT_IS_SYNCING` but it's never called during the login.

But the strange thing is that the login happens as you can see in this screenshot:
![image](https://user-images.githubusercontent.com/14869087/73055615-1c06c680-3e85-11ea-829f-52acb291af40.png)

Reply:
Congrats on figuring it out! 🙌 

Since that was the issue, can you test again with `'gi.actions/contract/sync'` and see if that has the same effect as `'gi.actions/contract/syncAndWait'`?